### ISSUE TITLE
Actively scrolling a live timeline results in scroll position jank

### ISSUE THREAD

**Issue by afontenot** [mastodon/mastodon#20552]
### Steps to reproduce the problem

1. Go to a live timeline like the federated timeline on mastodon.social.
2. Scroll down the page for a while, making slow scrolling movements on your touchpad / mouse wheel.
3. Scroll back up the page a little way as well.
...

### Expected behaviour

Both scrolling down the page and scrolling up the page are smooth, as if the timeline were one very long static sheet

### Actual behaviour

While scrolling is happening, if a new post gets loaded into the timeline, the scroll position suddenly jumps up or down

### Detailed description

Mastodon attempts to save the current scroll position when a live timeline is not at the top and jumps back to it after loading in a new post. However, when the user is actively scrolling the timeline, it seems not to update this scroll position quickly enough during ongoing scrolling. This means that if the timeline is updated during the scroll action, the saved position will be the *old* saved position, so if the user is scrolling down the page, they will get jumped back up when the new post gets loaded. And likewise, when scrolling up the page, they will get jumped back down.

It seems that the slower the browser is or the more posts you have loaded, the more jumping will occur. My guess is that JS just isn't firing the scroll events fast enough?

### Specifications

Mastodon 4.0.0rc3 on mastodon.social
Firefox 105.0.3

**Comment by gregmcdougall** [comment]
is this the same bug that causes the context menu from 'More' to slide up the screen in this view as new posts load?

**Comment by afontenot** [comment]
@gregmcdougall

No. But the scrolling / sliding / memory usage bugs (there are maybe half a dozen of them) could be fixed with a single change if the devs decided it was desirable. Basically, they need to stop adding new posts to the top of the timeline when the user has scrolled - that is to say, when the scroll position in the window is no no longer at the top. And then (if the timeline is not in slow mode) resume adding posts when the user scrolls back to the top. (Unless there are too many - more than 1000? - in which case it should offer the option to skip directly back to the top of the timeline.)

I don't know if that's desirable behavior. I have a user script that does this for me, it's just slightly too flaky to post publicly right now. Mastodon's timeline code has a race condition where if you add posts to the timeline too quickly, they can get added in the wrong order, fail to render, or even get "stuck" at the top of the timeline. The script tends to trigger this issue a lot. I haven't narrowed down the cause of this well enough to report it yet.

The specific bug you want to follow / thumb up for this is https://github.com/mastodon/mastodon/issues/14266. There are possible fixes for it that don't take the form I suggested above.

**Comment by mufarm** [comment]
I have a very similar problem with slight differences. Usually, when I reply to a post or click on a post, I will have a problem when I go back to the timeline, and when I try to scroll down, my focus will go all the way to the top. Then, when I scroll down, the page can get stuck on one post or go back and forth between two posts, and I'll have to select some text to scroll past those posts with the Pg Dn button. When I return to the timeline after responding or viewing a thread, if I can remember to select some text, then the page will scroll smoothly. Other people's new posts on the timeline don't seem to cause a problem.

The instance is running Mastodon v4.0.2+hometown-1.1.1.
